### PR TITLE
react nodes

### PR DESCRIPTION
**Pull request by rcraig12**
Hi, 

I hope these are a welcomed addition to the repo!

react installer : Install a react application at a specified file path
react dependency installer : Install dependencies for a react application using npm
dependency installer : internal node for react dependency installer.

any changes or additions please let me know

### CODE REVIEW

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      "735430f5-61ba-48c6-a8a6-e952edcb259e": {
+        "type": "string",
+        "name": "stderr"
+      },
+      "8b7a505c-873e-4936-9195-f1b3e28c5520": {
+        "type": "(string or null)",
+        "name": "error"
+      },
+      "4c3736af-6977-4632-b048-5446eaa87edd": {
+        "type": "any",
```
Comment (lvass74) [inline react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:36]: Type of `bounced` should be the same as the type of `sync`.
You can formulate it like this: 
```
typeof `sync`
```

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        "name": "bounced"
+      }
+    }
+  },
+  "name": "react/Dependency Installer",
```
Comment (lvass74) [inline react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:41]: If this is an internal node of  `react/React dependency installer` I suggest the conventional name of `react/React dependency installer/Internal`

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    "inputs": {
+      "1fc459fc-545e-4675-9916-b94bcab3563c": {
+        "type": "string[]",
+        "name": "dependencies",
+        "description": "Array of npm dependencies to install.\n\neg.\n\n[\"a\",\"b\"]"
+      },
+      "9484b15d-ca86-4d63-b07f-cb4f8f21e88b": {
+        "type": "string",
+        "name": "projectPath",
+        "description": "[Inherited from port `projectPath` of `syncer`] \nReceives full file path of react application."
```
Comment (lvass74) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:14]: Description should be revised.

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      },
+      "6668cb7f-874b-4c24-ae30-89fe2e24bad1": {
+        "type": "any",
+        "name": "start",
+        "description": "[Inherited from port `start` of `syncer`] \nReceives individual item for syncing."
```
Comment (lvass74) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:19]: Description should be revised.

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  },
+  "implementations": {
+    "cranq": {
+      "nodes": {
+        "1cd55979-d276-4251-8602-b1e7584641e6": {
+          "coordinates": [
+            95,
+            -55
+          ],
+          "name": "Dependency Installer",
```
Comment (lvass74) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:49]: By convention prototype names are nouns and start with capital letter whereas node (aka instance of prototypes) names are verbs (if possible) and start with lowercase letter.
The intentions is to have a kind of readable sentence describing the activity if we read the names of nodes following the connections.
In this case I'd suggest: `install dependencies`

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+          "prototype": "e4e1b2e6-26a1-4b3b-b472-4464814fa69a",
+          "params": {}
+        },
+        "0726521e-634a-461a-aa98-f4ea527b2bf8": {
+          "coordinates": [
+            -54,
+            14
+          ],
+          "name": "Syncer",
```
Comment (lvass74) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:58]: I suggest as name `sync parameters`

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 7]
```diff
@@ -0,0 +1,51 @@
+{
+  "implementations": {
+    "es6-node": {
+      "inputFunctions": {
+        "045a62ad-7340-479a-b544-f10d56f16597": "const { exec } = require( \"child_process\" );\nconst { sync } = data;\nconst { dependencies } = params;\nif (data.start !== false ){\n  const command = \"cd \" + data.projectPath + \" && \" + \"npm install \" + params.dependencies.join(' ');\n  outputs.stdout( command, tag );\n  if (state.running){\n    outputs.bounced(data.sync, tag);\n  } else {\n    state.running = true;\n    exec(command, (error, stdout, stderr) => {\n      if (error) {\n        outputs.error(`${error}`, tag);\n      } else {\n        outputs.stdout(stdout || \"\", tag);\n        outputs.stderr(stderr || \"\", tag);\n        outputs.error(null, tag);\n      }\n      state.running = false;\n    });\n  }\n}"
```
Comment (lvass74) [inline react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: The statement `const { sync } = data;` will always set the variable `sync` to undefined since data is supposed to contain "projectPath" and "start" as keys.

Comment (lvass74) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: The statement `outputs.bounced(data.sync, tag);` should be `outputs.bounced(data, tag);` since this is the input function of the port `sync` and it will get its value as `data`

Comment (lvass74) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: The parent node's `start` port is supposed to be a signal to start the operation. These kind of ports usually has `any` type and there's no need to validate their value (like here `if (data.start !== false )` since they are simple triggers, if any signal comes in - with whatever value - we start the operation, if we don't want to start, we should not trigger that port. Sending `false` to a start port is somewhat counterintuitive in data flow context imho.

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    "inputs": {
+      "df3107d7-179c-4e60-86b4-323813c9e86e": {
+        "type": "string",
+        "name": "projectPath",
+        "description": "Full file path of react application."
+      },
+      "32afb2d1-73c5-437c-a82a-8ec997b9a9c1": {
+        "type": "boolean",
+        "name": "replace",
+        "description": "Should the node delete and reinstall the application?\n\nboolean: defaults to false."
```
Comment (lvass74) [inline react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:19]: I'd add to the description that "Works as static parameter only." 
As it won't work if set by a signal arriving too late.

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    },
+    "outputs": {
+      "94074b45-8d35-46d1-8a73-f09c55487709": {
+        "type": "boolean",
+        "name": "trigger",
```
Comment (lvass74) [inline react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:25]: Maybe more descriptive name would be `success`

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      }
+    }
+  },
+  "name": "react/React Installer",
+  "iconId": "abac34e3-c832-4c7b-b045-62f82ebc3ddb",
+  "keywords": {
+    "react": null,
+    "installer": null
+  },
+  "description": "Installs a react application to a location set by string:projectPath",
```
Comment (lvass74) [inline react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:52]: Not clear what you mean by "string:projectPath".
By convention we use backtick to refer to a port or a node in the description.
In this case it would be:
```
Installs a react application to a location set by `projectPath`.
```

Comment (lvass74) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:52]: An example would be useful.
Note: if you use any path in a description always use "/" as path separator, not "\". Slash works on all platforms.

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json [hunk 11]
```diff
@@ -0,0 +1,56 @@
+{
+  "implementations": {
+    "es6-node": {
+      "inputFunctions": {
+        "df3107d7-179c-4e60-86b4-323813c9e86e": "const { exec } = require( \"child_process\" );\nconst { projectPath } = data;\nconst { replace } = params;\nlet command = \"npx create-react-app \" + data + \" --use-npm\";\nif (state.running){\n  outputs.bounced(data, tag);\n} else {\n  state.running = true;\n  if ( params.replace === true ) {\n    command = \"rm -rf \" + data + \" && npx create-react-app \" + data + \" --use-npm\";\n  }\n  outputs.stdout( command, tag );\n  exec(command, (error, stdout, stderr) => {\nif (error && (params.replace === false || 'undefined' )) {\n  outputs.error(`Application folder already exists. Set the replace input to true if you wish to reinstall the application from scratch`, tag);\n  outputs.trigger(false, tag);\n} else {\n  outputs.stdout(stdout || \"\", tag);\n  outputs.stderr(stderr || \"\", tag);\n  outputs.error(null, tag);\n  outputs.trigger(true, tag);\n}\n  state.running = false;\n  });\n}"
```
Comment (lvass74) [inline react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:5]: The statement `const { projectPath } = data;` is useless, as `data` is a string that contains the projectPath directly.

Comment (lvass74) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:5]: I think it is way too dangerous to run a command like this without any validation.
```
command = \"rm -rf \" + data + \" && npx create-react-app \" + data + \" --use-npm\";
```
I think we'd need to be sure the folder specified in `data` is really a folder and it really contains a react-app from a previous run.

Comment (lvass74) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:5]: `outputs.error(null, tag);`
No signal should be sent to an error port if there were no errors.
On the other hands if there were any the sent out value should be wrapped to this type at minimum: `{"error": string}`

**Review by lvass74 — CHANGES_REQUESTED** [review]
Thanks again for your efforts to contribute to Cranq!
Sorry for the amount of remarks, a lot of them is our faults not having published the prototype creation rules/conventions yet.
Let me know if you have any questions.

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json (hunk 11 above)
Comment (rcraig12) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:5]: > The statement `const { projectPath } = data;` is useless, as `data` is a string that contains the projectPath directly.

agree - const removed.

Comment (rcraig12) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:5]: > I think it is way too dangerous to run a command like this without any validation.
> 
> ```
> command = \"rm -rf \" + data + \" && npx create-react-app \" + data + \" --use-npm\";
> ```
> 
> I think we'd need to be sure the folder specified in `data` is really a folder and it really contains a react-app from a previous run.

agree - when the node creates a react app now it is additionally adding a file .cranq-react-app to make sure that is I want to delete the node before restarting it can check for the file as well as the path before deciding to remove. As an additional I realised the rm -rf was not of much use to win32 users so I have used fs.rmSync so that all os platforms will behave the same.

Comment (rcraig12) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:5]: > `outputs.error(null, tag);` No signal should be sent to an error port if there were no errors. On the other hands if there were any the sent out value should be wrapped to this type at minimum: `{"error": string}`

this is reused code from another node in the repository. I have now deleted the line. Additionally I have taken note of the errors being in typed to { "error" : "" } and have amended as necessary.

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json (hunk 9 above)
Comment (rcraig12) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:25]: renamed as suggested

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json (hunk 8 above)
Comment (rcraig12) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:19]: added - Use as a static parameter only.

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  "interface": {
+    "inputs": {
+      "93078915-107a-4854-baaf-e2cdd4c06760": {
+        "type": "string[]",
+        "name": "dependencies",
+        "description": "Array of npm dependencies to install.\n\neg.\n\n[\"a\",\"b\"]"
+      },
+      "045a62ad-7340-479a-b544-f10d56f16597": {
+        "type": "any",
```
[the remark this answers is not in the document]
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:17]: updated type to match description

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        "name": "sync",
+        "description": "{\n  \"projectPath\": string,\n  \"start\": boolean\n}"
```
[the remark this answers is not in the document]
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:19]: This is an internal node so description removed as it is `not reusable`

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  "interface": {
+    "inputs": {
+      "93078915-107a-4854-baaf-e2cdd4c06760": {
+        "type": "string[]",
+        "name": "dependencies",
+        "description": "Array of npm dependencies to install.\n\neg.\n\n[\"a\",\"b\"]"
+      },
+      "045a62ad-7340-479a-b544-f10d56f16597": {
+        "type": "any",
+        "name": "sync",
```
[the remark this answers is not in the document]
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:18]: renamed as suggested

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+      "86e1623a-d67c-4e74-98bb-449e29005d56": {
+        "type": "string",
+        "name": "stdout"
+      },
+      "735430f5-61ba-48c6-a8a6-e952edcb259e": {
+        "type": "string",
+        "name": "stderr"
+      },
+      "8b7a505c-873e-4936-9195-f1b3e28c5520": {
+        "type": "(string or null)",
```
[the remark this answers is not in the document]
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:32]: changed as suggested

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json (hunk 1 above)
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:36]: not sure need more details on what is meant?

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json (hunk 7 above)
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: > The statement `outputs.bounced(data.sync, tag);` should be `outputs.bounced(data, tag);` since this is the input function of the port `sync` and it will get its value as `data`

changed as suggested

Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: > The statement `const { sync } = data;` will always set the variable `sync` to undefined since data is supposed to contain "projectPath" and "start" as keys.

as per previous comment regarding React Installer line has been removed.

Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: > The parent node's `start` port is supposed to be a signal to start the operation. These kind of ports usually has `any` type and there's no need to validate their value (like here `if (data.start !== false )` since they are simple triggers, if any signal comes in - with whatever value - we start the operation, if we don't want to start, we should not trigger that port. Sending `false` to a start port is somewhat counterintuitive in data flow context imho.

I think the name start is probably confusing and creates the stated view so I will change it to `run` as I need to control the behaviour given the outcome is system calls.

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json (hunk 10 above)
Comment (rcraig12) [reply react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:52]: > Not clear what you mean by "string:projectPath". By convention we use backtick to refer to a port or a node in the description. In this case it would be:
> 
> ```
> Installs a react application to a location set by `projectPath`.
> ```

changed as per suggestion

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json (hunk 5 above)
Comment (rcraig12) [reply react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:49]: changed as suggested

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json (hunk 4 above)
Comment (rcraig12) [reply react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:19]: changed as identified

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json (hunk 3 above)
Comment (rcraig12) [reply react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:14]: changed as identified

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json (hunk 6 above)
Comment (rcraig12) [reply react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:58]: renamed as suggeted

File: react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json (hunk 1 above)
Comment (rcraig12) [reply react/dependency installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json]: reviewing the command runner node. It is using the same types as have used although the error output is now set to {"error": string} as suggested.

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      },
+      "9484b15d-ca86-4d63-b07f-cb4f8f21e88b": {
+        "type": "string",
+        "name": "projectPath",
+        "description": "Installation path for npm installer.\n"
+      },
+      "6668cb7f-874b-4c24-ae30-89fe2e24bad1": {
+        "type": "any",
+        "name": "run",
+        "description": "Run npm installer if boolean value received is true.\n"
```
Comment (danstocker) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:19]: Now that the type of `run` is `any`, it runs on any received value, not just true, and the description should reflect that.

File: react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  "interface": {
+    "inputs": {
+      "93078915-107a-4854-baaf-e2cdd4c06760": {
+        "type": "string[]",
+        "name": "dependencies",
+        "description": "Array of npm dependencies to install.\n\neg.\n\n[\"a\",\"b\"]"
+      },
+      "045a62ad-7340-479a-b544-f10d56f16597": {
+        "type": "{\n\"projectPath\": string, \n\"start\": boolean\n}",
```
Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:17]: The start property in this input seems irrelevant as it's only purpose is to convert the `projectPath` param into a signal.

File: react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 18]
```diff
@@ -0,0 +1,51 @@
+{
+  "implementations": {
+    "es6-node": {
+      "inputFunctions": {
+        "045a62ad-7340-479a-b544-f10d56f16597": "const { exec } = require( \"child_process\" );\nconst { dependencies } = params;\nif (data.start !== false ){\n  const command = \"cd \" + data.projectPath + \" && \" + \"npm install \" + params.dependencies.join(' ');\n  outputs.stdout( command, tag );\n  if (state.running){\n    outputs.bounced(data, tag);\n  } else {\n    state.running = true;\n    exec(command, (error, stdout, stderr) => {\n      if (error) {\n        outputs.error(`${error}`, tag);\n      } else {\n        outputs.stdout(stdout || \"\", tag);\n        outputs.stderr(stderr || \"\", tag);\n        outputs.error(null, tag);\n      }\n      state.running = false;\n    });\n  }\n}"
```
Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: Here `data.start !== false` will always be satisfied, (when used from the main dependency installer node) as the parent node's syncer adds a `run` property, not `start`. I think this condition is unnecessary anyway, as the value carried by start/run is `any`.

Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: Parameters `stdout` and `stderr` in the callback of `exec` can be `Buffer`s, so I'd serialize them the same way as `error` is serialized.
Also I would not send `null` on `error` when there was no error.

File: react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+      "735430f5-61ba-48c6-a8a6-e952edcb259e": {
+        "type": "any",
+        "name": "stderr"
+      },
+      "8b7a505c-873e-4936-9195-f1b3e28c5520": {
+        "type": "{ \"error\" : string}",
+        "name": "error"
+      },
+      "4c3736af-6977-4632-b048-5446eaa87edd": {
+        "type": "any",
```
Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:36]: Should be "typeof \`projectPath & start\`" as it simply forwards that input.

File: react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json (hunk 18 above)
Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:5]: On the callback error branch, we're sending a string instead of the documented `{"error": string}` which is the correct type. Should be something like: `{"error": String(error)}`, or `{"error": "Dependencies failed to install", "details": String(error)}`.

File: react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+      },
+      "045a62ad-7340-479a-b544-f10d56f16597": {
+        "type": "{\n\"projectPath\": string, \n\"start\": boolean\n}",
+        "name": "projectPath & start",
+        "description": ""
+      }
+    },
+    "outputs": {
+      "86e1623a-d67c-4e74-98bb-449e29005d56": {
+        "type": "any",
```
Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:24]: Should be `string` as we know that's what we send from the code.

File: react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+        "name": "stdout"
+      },
+      "735430f5-61ba-48c6-a8a6-e952edcb259e": {
+        "type": "any",
```
Comment (danstocker) [inline react/react dependency installer/installer/e4e1b2e6-26a1-4b3b-b472-4464814fa69a.json:28]: Should be `string` as we know that's what we send from the code.

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+      }
+    },
+    "outputs": {
```
Comment (danstocker) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:22]: Need to makes sure types reflect those of the internal code node. (see comments below)

File: react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json [hunk 23]
```diff
@@ -0,0 +1,127 @@
+{
+  "name": "react/React Dependency Installer",
+  "description": "Installs dependencies defined by the user for the react installer node. Node uses npm to install the dependencies on the same application path as the react application.",
+  "interface": {
+    "inputs": {
```
Comment (danstocker) [inline react/react dependency installer/8f8bd450-3f2b-45a5-abcc-12b633899c77.json:5]: We should point out in the descriptions that `dependencies` and `projectPath` are not protected from injection attack. (Or validate in code that they are actually path and package names)

File: react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json [hunk 24]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (danstocker) [inline react/react installer/e257d740-d47d-4794-82f1-3704850bdba9.json:1]: Most of the issues in the dependency installer repeat here, re. port types, serialization of `stdout`/`stderr`, etc.
Beyond that I feel the `success` output is redundant, as we either send something on `stdout` or `error`. If I need to turn that into a boolean, I can use a `boolean/Reverse condition` node.
One other thing I'd like to ask is to wrap this node into a structure node with an identical interface. As a general rule, we give preference to implementing nodes using existing nodes rather than JS, so this would keep the functionality of this node future-proof re. refactoring.

**Review by danstocker — CHANGES_REQUESTED** [review]
